theta_dup/include/theta_sketch_dup.h

```diff
@@ -0,0 +1,2209 @@
+  ptr += copy_from_mem(ptr, &flags_byte, sizeof(flags_byte));
+  uint16_t seed_hash;
+  ptr += copy_from_mem(ptr, &seed_hash, sizeof(seed_hash));
+
+  check_serial_version(serial_version, SERIAL_VERSION);
+  check_seed_hash(seed_hash, get_seed_hash(seed));
+
+  if (type == update_theta_sketch_dup_alloc<A>::SKETCH_TYPE) {
+    typename update_theta_sketch_dup_alloc<A>::resize_factor rf =
+        static_cast<typename update_theta_sketch_dup_alloc<A>::resize_factor>(
+            preamble_longs >> 6);
+    typedef typename std::allocator_traits<A>::template rebind_alloc<
+        update_theta_sketch_dup_alloc<A>>
+        AU;
+    return unique_ptr(
+        static_cast<theta_sketch_dup_alloc<A>*>(
+            new (AU().allocate(1)) update_theta_sketch_dup_alloc<A>(
+                update_theta_sketch_dup_alloc<A>::internal_deserialize(
+                    ptr, size - (ptr - static_cast<const char*>(bytes)), rf,
+                    lg_cur_size, lg_nom_size, flags_byte, seed))),
+        [](theta_sketch_dup_alloc<A>* ptr) {
+          ptr->~theta_sketch_dup_alloc();
+          AU().deallocate(static_cast<update_theta_sketch_dup_alloc<A>*>(ptr),
+                          1);
+        });
+  } else if (type == compact_theta_sketch_dup_alloc<A>::SKETCH_TYPE) {
+    typedef typename std::allocator_traits<A>::template rebind_alloc<
+        compact_theta_sketch_dup_alloc<A>>
+        AC;
+    return unique_ptr(
+        static_cast<theta_sketch_dup_alloc<A>*>(
+            new (AC().allocate(1)) compact_theta_sketch_dup_alloc<A>(
+                compact_theta_sketch_dup_alloc<A>::internal_deserialize(
+                    ptr, size - (ptr - static_cast<const char*>(bytes)),
+                    preamble_longs, flags_byte, seed_hash))),
+        [](theta_sketch_dup_alloc<A>* ptr) {
+          ptr->~theta_sketch_dup_alloc();
+          AC().deallocate(static_cast<compact_theta_sketch_dup_alloc<A>*>(ptr),
+                          1);
+        });
+  }
+  throw std::invalid_argument("unsupported sketch type " +
+                              std::to_string((int)type));
+}
+
+template <typename A>
+bool theta_sketch_dup_alloc<A>::isEqual(
+    const theta_sketch_dup_alloc<A>& r) const {
+  return this->theta_ == r.theta_ && this->is_empty_ == r.is_empty_;
+}
+
+template <typename A>
+uint16_t theta_sketch_dup_alloc<A>::get_seed_hash(uint64_t seed) {
+  HashState hashes;
+  MurmurHash3_x64_128(&seed, sizeof(seed), 0, hashes);
+  return hashes.h1;
+}
+
+template <typename A>
+void theta_sketch_dup_alloc<A>::check_sketch_type(uint8_t actual,
+                                                  uint8_t expected) {
+  if (actual != expected) {
+    throw std::invalid_argument("Sketch type mismatch: expected " +
+                                std::to_string((int)expected) + ", actual " +
+                                std::to_string((int)actual));
+  }
+}
+
+template <typename A>
+void theta_sketch_dup_alloc<A>::check_serial_version(uint8_t actual,
```

Comment:
comment for serial version?

Reply:
updated, see [this commit](https://github.com/googleinterns/cardinality-estimation-with-data-deletions/pull/2/commits/e34b8c6b92a191e115c621555d4316b1e5cf41b8)